File: solutions/day-09-design-hashset/Solution.cpp
#include <vector>
#include <list>

class MyHashSet {
private:
    static const int SIZE = 1000;
    std::vector<std::list<int>> buckets;

    int hash(int key) {
        return key % SIZE;
    }

public:
    MyHashSet() : buckets(SIZE) {}

    void add(int key) {
        auto& bucket = buckets[hash(key)];
        for (int val : bucket) {
            if (val == key) return;
        }
        bucket.push_back(key);
    }

    void remove(int key) {
        buckets[hash(key)].remove(key);
    }

    bool contains(int key) {
        for (int val : buckets[hash(key)]) {
            if (val == key) return true;
        }
        return false;
    }
};

File: solutions/day-09-design-hashset/Solution.java
import java.util.LinkedList;

class MyHashSet {

    private final int size = 1000;
    private LinkedList<Integer>[] buckets;

    public MyHashSet() {
        buckets = new LinkedList[size];
        for (int i = 0; i < size; i++) {
            buckets[i] = new LinkedList<>();
        }
    }

    private int hash(int key) {
        return key % size;
    }

    public void add(int key) {
        LinkedList<Integer> bucket = buckets[hash(key)];
        if (!bucket.contains(key)) {
            bucket.add(key);
        }
    }

    public void remove(int key) {
        buckets[hash(key)].remove((Integer) key);
    }

    public boolean contains(int key) {
        return buckets[hash(key)].contains(key);
    }
}
